### DISCUSSION TITLE
Serving media over RTMP

### DISCUSSION THREAD

**Discussion by calvin-at-spotify** [savonet/liquidsoap#3044] [Q&A] [answered]
Hi!

How can one configure liquidsoap to serve some media over RTMP? My understanding is `harbor.output` serves content over HTTP.

**Comment by toots** [comment]
This is possible using the `ffmpeg` backend. We have a `input.rtmp` wrapper included in the standard library:
```liquidsoap
# Read an RTMP stream.
# @category Source / Input
# @param ~max_buffer Maximum data buffer in seconds
# @param ~listen Act as a RTMP server and wait for incoming connection
# @param url URL to read RTMP from, in the form `rtmp://IP:PORT/ENDPOINT`
def input.rtmp(~id=null(), ~max_buffer=5., ~listen=true, url)
  input.ffmpeg(id=id, max_buffer=max_buffer, format="live_flv", self_sync=true,
               int_args=[("listen", listen ? 1 : 0)], url)
end
```

**Reply by toots** [reply, marked as answer]
Hmm sorry this is an input. An output would look like this:

```liquidsoap
s = single("/tmp/bla.mp4")

enc = %ffmpeg(
  format="flv",
  listen=1,
  %audio.copy,
  %video.copy
)

output.url(
  url="rtmp://localhost:1935/live/app",
  enc,
  s)
```

**Reply by calvin-at-spotify** [reply]
That would be *pushing* the content to `rtmp://localhost:1935/live/app`, wouldn't it? Is there a way to host `rtmp://localhost:1935/live/app` using liquidsoap?

**Reply by calvin-at-spotify** [reply]
This works as expected. If you want to expose it directly to the public internet, bind the output to `0.0.0.0`. i.e. `output.url(
  url="rtmp://0.0.0.0:1935/live/app"`

**Reply by calvin-at-spotify** [reply]
Moreover, you have to use encapsulate your data in `flv` frame: https://superuser.com/a/1659934
